ee/zentral/contrib/intune/api_client.py

```diff
@@ -0,0 +1,238 @@
+import logging
+import requests
+from asgiref.sync import async_to_sync
+
+from kiota_authentication_azure.azure_identity_authentication_provider import AzureIdentityAuthenticationProvider
+from azure.identity.aio import ClientSecretCredential
+from msgraph import GraphRequestAdapter, GraphServiceClient
+from datetime import datetime
+from zentral.contrib.inventory.conf import cleanup_windows_os_version, windows_version_from_build
+
+
+logger = logging.getLogger("zentral.contrib.intune.api_client")
+
+
+class Client:
+    scopes = ['https://graph.microsoft.com/.default']
+
+    def __init__(self, business_unit, tenant_id, client_id, client_secret):
+        self.business_unit = business_unit
+        self.tenant_id = tenant_id
+        self.client_id = client_id
+        self.client_secret = client_secret
+        self.session = requests.Session()
+        # Auth conf
+        self.auth_provider = AzureIdentityAuthenticationProvider(
+            ClientSecretCredential(
+                str(self.tenant_id),
+                str(self.client_id),
+                str(self.client_secret),
+            ),
+            scopes=self.scopes
+            )
+        self.request_adapter = GraphRequestAdapter(self.auth_provider)
+        self.graph_client = GraphServiceClient(self.request_adapter)
+
+    @classmethod
+    def from_tenant(cls, tenant):
+        ''' Setting un credentials and the Graph API Client'''
+        client = cls(
+            tenant.business_unit,
+            tenant.tenant_id,
+            tenant.client_id,
+            tenant.get_client_secret(),
+        )
+        return client
+
+    def iter_machine_snapshot_trees(self):
+        # TODO: Use batches for processing
+        # https://learn.microsoft.com/en-us/graph/query-parameters?tabs=http
+        results = []
+
+        @async_to_sync
+        async def consume_data():
+            async for device in self.iter_devices():
+                results.append(device)
+        consume_data()
+
+        # FIXME: get a batch of devices to avoid a big results array
+        for device in results:
+            try:
+                # TODO check last_sync_date_time
+                yield self.build_machine_snapshot_tree(device)
+            except Exception:
+                logger.exception("Device %s: could not build machine snapshot tree", device.id)
+
+    async def iter_devices(self):
+        ''' TODO:
+            - Use device_batch_size to retrieve results
+            - Use last_sync_date_time to limit results
```

Comment:
better to order by device id, and use the "top" query parameter, and device id > last seen device id if possible. Only a suggestion, didn't look really deep in the docs.

Reply:
I'll modify the docstring.